ddht/tools/lru_sql_dict.py

```diff
@@ -0,0 +1,269 @@
+            nref = Node(*nref_node)
+            self._execute(
+                "UPDATE cache SET nref=? WHERE key=?;", (None, nref.key),
+            )
+        elif pref_node:
+            pref = Node(*pref_node)
+            self._execute(
+                "UPDATE cache SET pref=? WHERE key=?;", (None, pref.key),
+            )
+
+    def _insert_item(self, key: TKey, value: TValue) -> None:
+        # insert new item into map and cache
+        serialized_key = self.key_encoder(key)
+        serialized_value = self.value_encoder(value)
+        if self.is_empty:
+            new_pref = None
+            new_nref = None
+
+            self._execute(
+                "INSERT INTO cache VALUES (?, ?, ?, ?);",
+                (serialized_key, serialized_value, new_pref, new_nref),
+            )
+
+        else:
+            # evict least recently used key/value
+            if self.is_full:
+                self.__delitem__(self.key_decoder(self.tail.key))
+
+            # get old head
+            old_head_key, _, old_head_pref, _ = self.head
+
+            # add new head to cache
+            self._execute(
+                "INSERT INTO cache VALUES (?, ?, ?, ?);",
+                (serialized_key, serialized_value, None, old_head_key),
+            )
+
+            # update old head in cache
+            self._execute(
+                "UPDATE cache SET pref=? WHERE key=?;", (serialized_key, old_head_key),
+            )
+
+    def _update_item(self, key: TKey, value: TValue) -> None:
+        serialized_key = self.key_encoder(key)
+        serialized_value = self.value_encoder(value)
+        self._execute(
+            "UPDATE cache SET value=? WHERE key=?;", (serialized_value, serialized_key)
+        )
+
+    def _fetch_single_query(self, query: str, args: Tuple[Any, ...] = None) -> Any:
+        with self._conn:
+            if args:
+                cursor = self._conn.execute(query, args).fetchall()
+            else:
+                cursor = self._conn.execute(query).fetchall()
+        if len(cursor) > 1:
+            raise Exception(
+                f"Invalid db state. More than one result found for query: {query}."
+            )
+        if not cursor:
+            return None
+        return first(cursor)
+
+    def _execute(self, query: str, args: Tuple[Any, ...] = None) -> None:
+        with self._conn:
+            if args:
+                self._conn.execute(query, args)
+            else:
+                self._conn.execute(query)
+
+    @property
+    def is_full(self) -> bool:
+        if not self.cache_size:
+            return False
+        return self.__len__() >= self.cache_size
+
+    @property
+    def is_empty(self) -> bool:
+        return self.__len__() == 0
+
+    @property
+    def head(self) -> Node:
+        head = self._fetch_single_query(
+            "SELECT key,value,pref,nref FROM cache WHERE pref IS NULL;"
+        )
+        if not head:
+            raise Exception("No head found.")
```

Comment:
I'm not 100% on this being the right behavior, but `KeyError` or `IndexError` seem appropriate.  Conceptually this is close to `dict.pop()`.